examples/ct/ct_handlers.go

```diff
@@ -436,3 +522,86 @@ func marshalAndWriteAddChainResponse(sct ct.SignedCertificateTimestamp, km crypt
 
 	return nil
 }
+
+func parseAndValidateGetEntriesRange(r *http.Request, maxAllowedRange int64) (int64, int64, error) {
+	startIndex, err := strconv.ParseInt(r.FormValue(getEntriesParamStart), 10, 64)
+
+	if err != nil {
+		return 0, 0, err
+	}
+
+	endIndex, err := strconv.ParseInt(r.FormValue(getEntriesParamEnd), 10, 64)
+
+	if err != nil {
+		return 0, 0, err
+	}
+
+	return validateStartAndEnd(startIndex, endIndex, maxAllowedRange)
+}
+
+// validateStartAndEnd applies validation to the range params for get-entries. Returns nil if
+// we accept them or an error giving the reason why not
+func validateStartAndEnd(start, end, maxRange int64) (int64, int64, error) {
+	if start < 0 || end < 0 {
+		return 0, 0, fmt.Errorf("start (%d) and end (%d) parameters must be >= 0", start, end)
+	}
+
+	if start > end {
+		return 0, 0, fmt.Errorf("start (%d) and end (%d) is not a valid range", start, end)
+	}
+
+	numEntries := end - start + 1
+
+	if numEntries > maxRange {
+		return 0, 0, fmt.Errorf("requesting %d entries but we only allow up to %d", numEntries, maxRange)
+	}
+
+	return start, end, nil
+}
+
+// buildIndicesForRange expands the range out, the backend allows for non contiguous leaf fetches
+// but the CT spec doesn't. The input values should have been checked for consistency before calling
+// this.
+func buildIndicesForRange(start, end int64) []int64 {
+	indices := make([]int64, 0, end-start+1)
+	for i := start; i <= end; i++ {
+		indices = append(indices, i)
+	}
+
+	return indices
+}
+
+// isResponseContiguousRange checks that the response has a contiguous range of leaves and
+// that it is a subset or equal to the requested range. This is additional protection against
+// backend bugs. Returns nil if the response looks valid.
+func isResponseContiguousRange(response *trillian.GetLeavesByIndexResponse, start, end int64) error {
+	for li, l := range response.Leaves {
+		if li > 0 && response.Leaves[li].LeafIndex-response.Leaves[li-1].LeafIndex != 1 {
+			return fmt.Errorf("backend returned non contiguous leaves: %v %v", response.Leaves[li-1], response.Leaves[li])
+		}
+
+		if l.LeafIndex < start || l.LeafIndex > end {
+			return fmt.Errorf("backend returned leaf:%d outside requested range:%d, %d", l.LeafIndex, start, end)
+		}
+	}
+
+	return nil
+}
+
+// marshalGetEntriesResponse does the conversion from the backend response to the one we need for
+// an RFC compliant JSON response to the client.
+func marshalGetEntriesResponse(rpcResponse *trillian.GetLeavesByIndexResponse) (getEntriesResponse, error) {
+	jsonResponse := getEntriesResponse{}
+
+	for _, leaf := range rpcResponse.Leaves {
+		// We're only deserializing it to ensure it's valid, don't need the result
```

Comment:
I think I want to leave this check in for now so it can be hooked up with monitoring later as this probably indicates data corruption. I can make it return the bad data if you think that's better.